x-pack/legacy/plugins/snapshot_restore/server/routes/api/app.ts

```diff
@@ -19,43 +23,72 @@ export function getXpackMainPlugin() {
   return xpackMainPlugin;
 }
 
-export const getPermissionsHandler: RouterRouteHandler = async (req, callWithRequest) => {
+const extractMissingPrivileges = (privilegesObject: { [key: string]: boolean }): string[] => {
+  return Object.keys(privilegesObject).reduce(
+    (privileges: string[], privilegeName: string): string[] => {
+      if (!privilegesObject[privilegeName]) {
+        privileges.push(privilegeName);
+      }
+      return privileges;
+    },
+    []
+  );
+};
+
+export const getPermissionsHandler: RouterRouteHandler = async (
+  req,
+  callWithRequest
+): Promise<AppPermissions> => {
   const xpackInfo = getXpackMainPlugin() && getXpackMainPlugin().info;
   if (!xpackInfo) {
     // xpackInfo is updated via poll, so it may not be available until polling has begun.
     // In this rare situation, tell the client the service is temporarily unavailable.
     throw wrapCustomError(new Error('Security info unavailable'), 503);
   }
 
+  const permissionsResult: AppPermissions = {
+    hasPermission: true,
+    missingClusterPrivileges: [],
+    missingIndexPrivileges: [],
+  };
+
   const securityInfo = xpackInfo && xpackInfo.isAvailable() && xpackInfo.feature('security');
   if (!securityInfo || !securityInfo.isAvailable() || !securityInfo.isEnabled()) {
     // If security isn't enabled, let the user use app.
-    return {
-      hasPermission: true,
-      missingClusterPrivileges: [],
-    };
+    return permissionsResult;
   }
 
+  // Get cluster priviliges
   const { has_all_requested: hasPermission, cluster } = await callWithRequest('transport.request', {
     path: '/_security/user/_has_privileges',
     method: 'POST',
     body: {
-      cluster: APP_PERMISSIONS,
+      cluster: APP_REQUIRED_CLUSTER_PRIVILAGES,
     },
   });
 
-  const missingClusterPrivileges = Object.keys(cluster).reduce(
-    (permissions: string[], permissionName: string): string[] => {
-      if (!cluster[permissionName]) {
-        permissions.push(permissionName);
-      }
-      return permissions;
-    },
-    []
-  );
+  // Find missing cluster privileges and set overall app permissions
+  permissionsResult.missingClusterPrivileges = extractMissingPrivileges(cluster || {});
+  permissionsResult.hasPermission = hasPermission;
 
-  return {
-    hasPermission,
-    missingClusterPrivileges,
-  };
+  // Get all index privileges the user has
+  const { indices } = await callWithRequest('transport.request', {
+    path: '/_security/user/_privileges',
+    method: 'GET',
+  });
+
+  // Check if they have at the required index privilages for at least one index
+  const oneIndexWithAllPrivileges = indices.find(({ privileges }: { privileges: string[] }) => {
+    return (
+      privileges.includes('all') ||
+      APP_RESTORE_INDEX_PRIVILAGES.filter(privilage => !privileges.includes(privilage)).length === 0
```

Comment:
I had to read through this a few times to figure out the logic. Could we extract this condition and add a variable to communicate meaning? Here's my suggestion, it's a couple more lines but I find it easier to understand. (Warning, I haven't tested it!)

```js
  // Check if they have all the required index privilages for at least one index
  const hasIndexWithRestorePrivileges = indices.find(({ privileges }: { privileges: string[] }) => {
    if (privileges.includes('all')) {
      return true;
    }
    
    const indexHasRestorePrivileges = APP_RESTORE_INDEX_PRIVILEGES.every(restorePrivilege => privileges.includes(restorePrivilege));
    return indexHasRestorePrivileges;
  });
```

I know we could eliminate a line by getting rid of the `indexHasRestorePrivileges` variable, but its name makes it so much easier for me to follow the logic. :)